File: debop4s-data/src/main/scala/com/github/debop4s/data/hibernate/tools/StatelessTool.scala
package com.github.debop4s.data.hibernate.tools

import org.hibernate.{Session, HibernateException, SessionFactory, StatelessSession}
import org.slf4j.LoggerFactory
import scala.annotation.varargs

/**
 * Hibernate [[StatelessSession]] 을 이용한 작업을 수행합니다.
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2014. 2. 25. 오후 5:35
 */
object StatelessTool {

    private lazy val log = LoggerFactory.getLogger(getClass)

    def openStatelessSession(session: Session): StatelessSession =
        session.getSessionFactory.openStatelessSession()

    def executeTransactional(sessionFactory: SessionFactory)(action: StatelessSession => Unit) {
        log.debug(s"StatelessSession을 이용하여 transaction 작업을 수행합니다...")

        val stateless = sessionFactory.openStatelessSession()
        val tx = stateless.beginTransaction()

        try {
            action(stateless)
            tx.commit()
        } catch {
            case e: Throwable =>
                log.error(s"StatelessSession을 이용한 작업에 실패했습니다. rollback합니다.", e)
                if (tx != null) tx.rollback()
                throw new HibernateException(e)
        } finally {
            if (stateless != null)
                stateless.close()
        }
    }

    @varargs
    def executeAllTransactional(sessionFactory: SessionFactory, actions: (StatelessSession => Unit)*) {
        executeAllTransactional(sessionFactory, actions.toIterable)
    }

    def executeAllTransactional(sessionFactory: SessionFactory, actions: Iterable[StatelessSession => Unit]) {
        log.debug(s"StatelessSession을 이용하여 transaction 작업을 수행합니다...")

        val stateless = sessionFactory.openStatelessSession()
        val tx = stateless.beginTransaction()

        try {
            actions.foreach(action => action(stateless))
            tx.commit()
        } catch {
            case e: Throwable =>
                log.error(s"StatelessSession을 이용한 작업에 실패했습니다. rollback합니다.", e)
                if (tx != null) tx.rollback()
                throw new HibernateException(e)
        } finally {
            if (stateless != null)
                stateless.close()
        }
    }

    def execute(sessionFactory: SessionFactory, action: StatelessSession => Unit) {
        val stateless = sessionFactory.openStatelessSession()

        try {
            action(stateless)
        } catch {
            case e: Throwable =>
                log.error(s"StatelessSession을 이용한 작업에 실패했습니다.", e)
                throw new HibernateException(e)
        } finally {
            if (stateless != null)
                stateless.close()
        }
    }

    @varargs
    def executeAll(sessionFactory: SessionFactory, actions: (StatelessSession => Unit)*) {
        executeAll(sessionFactory, actions.toIterable)
    }

    def executeAll(sessionFactory: SessionFactory, actions: Iterable[StatelessSession => Unit]) {
        log.debug(s"StatelessSession을 이용하여 작업을 수행합니다...")

        val stateless = sessionFactory.openStatelessSession()

        try {
            actions.foreach(action => action(stateless))
        } catch {
            case e: Throwable =>
                log.error(s"StatelessSession을 이용한 작업에 실패했습니다.", e)
                throw new HibernateException(e)
        } finally {
            if (stateless != null)
                stateless.close()
        }
    }
}

File: debop4s-data-orm/src/main/scala/debop4s/data/orm/hibernate/tools/StatelessTool.java
package debop4s.data.orm.hibernate.tools;

import debop4s.core.JAction1;
import lombok.extern.slf4j.Slf4j;
import org.hibernate.Session;
import org.hibernate.SessionFactory;
import org.hibernate.StatelessSession;
import org.hibernate.Transaction;

/**
 * {@link StatelessSession} 을 이용한 작업을 수행합니다.
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 11. 8. 오후 7:26
 * @deprecated use {@link debop4s.data.orm.hibernate.utils.StatelessUtils}
 */
@Deprecated
@Slf4j
public final class StatelessTool {

    private StatelessTool() { }

    public static StatelessSession openStatelessSession(Session session) {
        return session.getSessionFactory().openStatelessSession();
    }

    public static void executeTransactional(SessionFactory sessionFactory, JAction1<StatelessSession> action) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = sessionFactory.openStatelessSession();
        Transaction tx = stateless.beginTransaction();
        try {
            action.perform(stateless);
            tx.commit();
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            if (tx != null) tx.rollback();
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }

    public static void executeTransactional(Session session, JAction1<StatelessSession> action) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = openStatelessSession(session);
        Transaction tx = stateless.beginTransaction();
        try {
            action.perform(stateless);
            tx.commit();
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            if (tx != null) tx.rollback();
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }

    @SafeVarargs
    public static void executeTransactional(Session session, JAction1<StatelessSession>... actions) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = openStatelessSession(session);
        Transaction tx = stateless.beginTransaction();

        try {
            for (JAction1<StatelessSession> action : actions) {
                action.perform(stateless);
            }
            tx.commit();
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            if (tx != null) tx.rollback();
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }

    public static void executeTransactional(Session session, Iterable<JAction1<StatelessSession>> actions) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = openStatelessSession(session);
        Transaction tx = stateless.beginTransaction();

        try {
            for (JAction1<StatelessSession> action : actions) {
                action.perform(stateless);
            }
            tx.commit();
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            if (tx != null) tx.rollback();
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }


    public static void execute(SessionFactory sessionFactory, JAction1<StatelessSession> action) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = sessionFactory.openStatelessSession();
        try {
            action.perform(stateless);
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }

    public static void execute(Session session, JAction1<StatelessSession> action) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = openStatelessSession(session);
        try {
            action.perform(stateless);
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }

    @SafeVarargs
    public static void execute(Session session, JAction1<StatelessSession>... actions) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = openStatelessSession(session);

        try {
            for (JAction1<StatelessSession> action : actions) {
                action.perform(stateless);
            }
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }

    public static void execute(Session session, Iterable<JAction1<StatelessSession>> actions) {
        log.debug("StatelessSession을 이용하여 Transaction 작업을 수행합니다...");

        StatelessSession stateless = openStatelessSession(session);

        try {
            for (JAction1<StatelessSession> action : actions) {
                action.perform(stateless);
            }
        } catch (Exception e) {
            log.error("StatelessSession을 이용한 작업에 실패했습니다. rollback 합니다.", e);
            throw new RuntimeException(e);
        } finally {
            if (stateless != null)
                stateless.close();
        }
    }


}
